Hiding dashboard cards: add Personalize Home Tab screen

kean opened the pull request:
Related issue: #20296

Previous PRs (dependencies):

1. #20365

This PR introduces a new "Personalize Home Tab" button to the Home tab (Jetpack-only). It opens [a new screen](https://jetpackmobile.wordpress.com/2023/03/14/hiding-dashboard-cards-i2-ios/) that allows toggling visibility for _most_ of the cards on the dashboard.

> Not every card is made personalizable yet (see `DashboardCard.personalizableCards` that controls the list). The remaining cards require special attention and will be covered separately.

## Testing Instructions

*Prerequisites*
- "Personalize Home Tab" feature flag is enabled

**Test Case 1**
- Select site A and scroll to the bottom of the Home tab
- Tap "Personalize Home Tab"
- Use the switches to hide one of the cards
- Verify that the card got hidden on the Home tab
- Switch to site B and open the Home tab
- Verify that the previously hidden card is still visible (preferences are saved on a per-site basis)

_Testing Notes_
- Test on iOS 14 and other version to make sure the design is consistent across platforms and everything works as expected
- Test on iPad
- Try different font sizes (dynamic type), themes (light/dark), RTL languages, VoiceOver

## Screenshots

| Add button to Home tab      | Add Personalize screen |
| ----------- | ----------- |
|![Simulator Screen Shot - iPhone 14 Pro - 2023-03-20 at 19 00 22](https://user-images.githubusercontent.com/1567433/226484580-45db9da1-3b4d-4671-8d8c-a8a7947fbd23.png)| ![Simulator Screen Shot - iPhone 14 Pro - 2023-03-20 at 19 00 08](https://user-images.githubusercontent.com/1567433/226484591-f0389ad5-5a6e-4091-8d93-0d114de01182.png) |

| Dynamic Type      | Dynamic Type |
| ----------- | ----------- |
| ![Simulator Screen Shot - iPhone 14 Pro - 2023-03-20 at 19 00 22](https://user-images.githubusercontent.com/1567433/226484855-76205485-057e-432f-8e93-cfcfca23a2b5.png) | ![Simulator Screen Shot - iPhone 14 Pro - 2023-03-20 at 19 00 14](https://user-images.githubusercontent.com/1567433/226484813-79203330-cadd-4aa2-9789-65eb8957ca09.png) |

| Dark Mode      | Dark Mode |
| ----------- | ----------- |
|![Simulator Screen Shot - iPhone 14 Pro - 2023-03-20 at 20 15 15](https://user-images.githubusercontent.com/1567433/226492286-35ee3dce-5843-48cb-86a8-3750bceee641.png) |![Simulator Screen Shot - iPhone 14 Pro - 2023-03-20 at 20 15 19](https://user-images.githubusercontent.com/1567433/226492309-20cc2615-e134-42db-bc6d-8de7494a3bca.png) 

| RTL      | RTL |
| ----------- | ----------- |
| ![Simulator Screen Shot - iPhone 14 Pro - 2023-03-20 at 18 59 20](https://user-images.githubusercontent.com/1567433/226484911-7a33729a-ea95-4ced-8e71-281ed025c63f.png) | ![Simulator Screen Shot - iPhone 14 Pro - 2023-03-20 at 18 59 26](https://user-images.githubusercontent.com/1567433/226484921-983176bf-13d7-4e0a-bf6a-a8824bd277bb.png) |

|iPad|
| ----------- | 
| ![Screenshot 2023-03-27 at 3 50 13 PM](https://user-images.githubusercontent.com/1567433/228051768-0fad803c-789d-4815-a838-c6a75b92f5e7.png) |

## Regression Notes
1. Potential unintended areas of impact: **n/a** (it's a new screen, so I don't expect it to affect any of the existing features)
3. What I did to test those areas of impact (or what existing automated tests I relied on): **n/a**
4. What automated tests I added (or what prevented me from doing so): I added a few unit tests for the new screen

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

kean committed:
Add Personalize Home Tab screen

kean commented on WordPress/Classes/Utility/BuildInformation/FeatureFlag.swift:
```diff
@@ -155,6 +156,8 @@ enum FeatureFlag: Int, CaseIterable, OverrideableFlag {
         case .readerUserBlocking:
             return true
+        case .personalizeHomeTab:
+            return false
```
When is it going to be a good time to enable it?

guarani replied:
It's normally done in the last PR of the project.

guarani replied:
During development, you can enable these via an in-app menu in app settings.

kean replied:
Thanks for the clarification. It'd be nice to have a way to override it during the build process, for example, using a local configuration file added to gitignore.

kean committed:
Fix BlogDashboardPersonalizationView design on iOS 14

kean committed:
Add analyticts to new personalization flow on dashboard

kean committed:
Add BlogDashboardPersonalizationView iPad support

kean committed:
Add BlogDashboardPersonalizationViewModel tests and docs

kean committed:
Update to support latest BlogDashboardPersonalizationService changes

guarani commented:
👋 @kean, is there a good way to see just the changes that this PR introduces without requiring https://github.com/wordpress-mobile/WordPress-iOS/pull/20365 to be merged first?

kean commented:
> 👋 @kean, is there a good way to see just the changes that this PR introduces without requiring #20365 to be merged first?

Yes, @guarani, this PR has the following commits: https://github.com/wordpress-mobile/WordPress-iOS/pull/20369/files/2e08a5ec31734e3ea60a786aea748866c0b79638..1db57f64dac8911765444ab7de959ce79a429c19.

guarani commented on RELEASE-NOTES.txt:
```diff
@@ -1,6 +1,6 @@
 22.1
 -----
-
+* [*] Add a Personalize Home Tab button to the bottom of the Home tab that allows changing visibility for the cards on the Home tab [#20369]
```
Nice job on the release notes. You could change this to `**` to give this a signal this is of higher importance when   the final editorialized release notes are made. However, I know this isn't documented publically, so I've created a PR to address this: https://github.com/wordpress-mobile/WordPress-iOS/pull/20402.

kean replied:
I increased the priority (and also fixes the merge conflicts in the release notes).

guarani commented on WordPress/Classes/ViewRelated/Blog/BlogPersonalization/BlogDashboardPersonalizationView.swift:
```diff
@@ -0,0 +1,59 @@
+    struct Strings {
+        static let title = NSLocalizedString("personalizeHome.title", value: "Personalize Home Tab", comment: "Page title")
+        static let sectionHeader = NSLocalizedString("personalizeHome.cardsSectionHeader", value: "Add or hide cards", comment: "Section header")
+        static let sectionFooter = NSLocalizedString("personalizeHome.cardsSectionFooter", value: "Cards may show different content depending on what's happening on your site. We're working on on more cards and controls.", comment: "Section footer displayed below the list of toggles")
```
```suggestion
        static let sectionFooter = NSLocalizedString("personalizeHome.cardsSectionFooter", value: "Cards may show different content depending on what's happening on your site. We're working on more cards and controls.", comment: "Section footer displayed below the list of toggles")
```

guarani commented on WordPress/Classes/ViewRelated/Blog/Blog Dashboard/Cards/BlogDashboardPersonalizeCardCell.swift:
```diff
@@ -0,0 +1,90 @@
+
+    // MARK: - View setup
+
+    private func setupView() {
+        let titleLabel = UILabel()
+        titleLabel.text = Strings.buttonTitle
+        titleLabel.font = WPStyleGuide.fontForTextStyle(.subheadline, fontWeight: .semibold)
+        titleLabel.adjustsFontForContentSizeCategory = true
+
+        let imageView = UIImageView(image: UIImage(named: "personalize")?.withRenderingMode(.alwaysTemplate))
+        imageView.tintColor = .label
+
+        let spacer = UIView()
+        spacer.translatesAutoresizingMaskIntoConstraints = false
+        spacer.widthAnchor.constraint(greaterThanOrEqualToConstant: 8).isActive = true
+
+        let contents = UIStackView(arrangedSubviews: [titleLabel, spacer, imageView])
+        contents.alignment = .center
+        contents.isUserInteractionEnabled = false
+
+        personalizeButton.accessibilityLabel = Strings.buttonTitle
+        personalizeButton.setBackgroundImage(.renderBackgroundImage(fill: .tertiarySystemFill), for: .normal)
+        personalizeButton.addTarget(self, action: #selector(buttonTapped), for: .touchUpInside)
+
+        let container = UIView()
+        container.layer.cornerRadius = 10
+        container.addSubview(personalizeButton)
+        container.addSubview(contents)
+
+        personalizeButton.translatesAutoresizingMaskIntoConstraints = false
+        container.pinSubviewToAllEdges(personalizeButton)
+
+        contents.translatesAutoresizingMaskIntoConstraints = false
+        container.pinSubviewToAllEdges(contents, insets: .init(allEdges: 16))
+
+        contentView.addSubview(container)
+        container.translatesAutoresizingMaskIntoConstraints = false
+        contentView.pinSubviewToAllEdges(container)
+    }
```
not a blocker: Could a `UIButton`'s text and images properties and insets simplify this by avoiding separate subviews? I'm curious about your thoughts here.

kean replied:
Making `UIButton` look like this isn't easy. By default, it displays the image on the left side, and there is no (good) way to set flexible spacing for the image. I wish it was as easy to use as the SwiftUI `Button`.

There is a new `UIButton.Configuration` that almost gets you there, but it's iOS 15+ only.

<img width="971" alt="Screenshot 2023-03-24 at 9 54 47 PM" src="https://user-images.githubusercontent.com/1567433/227679875-ea8fc90c-4cd8-4dec-91cc-ef66d37a50cd.png">

There is a way to do it with negative image/title insets, but then again, there is flexible spacing. I could use `layoutSubviews` to adjust the spacing dynamically based on the container and the title size. But it would've been over-complicated.

I considered using a simple gesture recognizer for handling taps on this card, but `UIButton` has a nice highlighted state and works better for accessibility.

guarani replied:
Thanks for sharing! I agree about `UIButton` image/title insets being complicated.

guarani reviewed (approved):
The new personalization is looking great! 

I gave this a test and things worked smoothly except for a visual glitch when hiding all cards while the screen is scrolled all the way to the bottom:

https://user-images.githubusercontent.com/1898325/227656711-fe8057d5-b7ce-4183-8e35-b97c44f6bdb3.mov

Trying this out, I can see how the personalization options for the "draft post" card and "scheduled post" dashboard cards can be unintuitive when there is no draft or scheduled post. Perhaps this needs to be made more intuitive before we release this. I don't think this is a blocker to this PR, but worth addressing before we ship this to users.

I'm approving this but I think we need @osullivanchris's input to give this. Some installable builds should pop up here, I've run CI now and it's building as of the time of writing.

guarani commented:
> Yes, @guarani, this PR has the following commits: https://github.com/wordpress-mobile/WordPress-iOS/pull/20369/files/2e08a5ec31734e3ea60a786aea748866c0b79638..1db57f64dac8911765444ab7de959ce79a429c19.

That works and is an option, I did that here now to review the PR. What do you think about targeting the previous PR's branch (the one this depends on) instead of `trunk`? A nice effect is that when the first PR is merged, GitHub should automatically update this PR to point to `trunk`, so it shouldn't add manual work. Just a suggestion, and not sure if you've already considered it.

kean commented:
> What do you think about targeting the previous PR's branch (the one this depends on) instead of trunk?

Great thinking, but It doesn't seem to be possible for branches in forks. The branch from the previous PR doesn't exist in the main repo and I can't target it.

> I gave this a test and things worked smoothly except for a visual glitch when hiding all cards while the screen is scrolled all the way to the bottom:

Nice catch. I put it on the list of defects.

kean committed:
Fix localizable string value

Co-authored-by: Paul Von Schrottky <paulvonschrottky@gmail.com>

osullivanchris commented:
> Trying this out, I can see how the personalization options for the "draft post" card and "scheduled post" dashboard cards can be unintuitive when there is no draft or scheduled post. Perhaps this needs to be made more intuitive before we release this. I don't think this is a blocker to this PR, but worth addressing before we ship this to users.

> I'm approving this but I think we need @osullivanchris's input to give this. Some installable builds should pop up here, I've run CI now and it's building as of the time of writing.

Oh super interesting, good catch @guarani 

I think there's two problems: scroll position after personalising, and turning on cards with no content.

Re; scroll position. What do you both think about this. If no changes made, retain scroll position. If user taps some switches, scroll back to top when the modal presentation closes. Does that make sense?

Re; draft/scheduled posts. I have that line at the bottom of the list saying that some cards won't have content. But it could be easily ignored in this case. I can think of two things:
1. We could add subheadings to the list items, describing the card a little. E.g. title 'Draft Posts', subtitle 'Shows any recently created/edited drafts'. 
2. We could add subheadings only to items that need some warning or description e.g. title: 'Draft posts', caption 'Only shows if you have recently created/edited draft posts'
3. We could show some empty state for the card on the dashboard. I don't really love this though.

Any opinions on those 3 approaches?

kean commented:
> Re; scroll position. What do you both think about this. If no changes made, retain scroll position. If user taps some switches, scroll back to top when the modal presentation closes. Does that make sense?

The behavior from [the video](https://github.com/wordpress-mobile/WordPress-iOS/pull/20369#pullrequestreview-1357343984) is a defect (minor). I added it to my list and will address it in one of the upcoming PRs. The only way to access the Personalize button is by scrolling to the very bottom. I think it'll be best if, after you close the Personalize screen, you'll be back to where you were: at the bottom of the screen.

> Re; draft/scheduled posts. I have that line at the bottom of the list saying that some cards won't have content. But it could be easily ignored in this case.

It depends on how you think about these widgets. If you take the iOS Today view, you, as a user, are expected to create your own dashboard by adding widgets. If you add a widget, and it doesn't show up, it would've been weird. But in the case of Wordpress, all widgets are already enabled by default. 

I think of the Personalize screen in its current state as either a way to undo the "Hide this" action or as a way to filter out the widgets you don't want to see. The section footer text at the bottom already alludes to the fact that the cards are dynamic: "Cards may show different content depending on what's happening on your site."

kean committed:
Update personalize screen size on iPad

osullivanchris commented:
> The behavior from https://github.com/wordpress-mobile/WordPress-iOS/pull/20369#pullrequestreview-1357343984 is a defect (minor). I added it to my list and will address it in one of the upcoming PRs. The only way to access the Personalize button is by scrolling to the very bottom. I think it'll be best if, after you close the Personalize screen, you'll be back to where you were: at the bottom of the screen.

Sounds fine to start with. 


> It depends on how you think about these widgets. If you take the iOS Today view, you, as a user, are expected to create your own dashboard by adding widgets. If you add a widget, and it doesn't show up, it would've been weird. But in the case of Wordpress, all widgets are already enabled by default.

> I think of the Personalize screen in its current state as either a way to undo the "Hide this" action or as a way to filter out the widgets you don't want to see. The section footer text at the bottom already alludes to the fact that the cards are dynamic: "Cards may show different content depending on what's happening on your site."

You've made a good case there, thanks. Let's stick with what we have.

guarani commented:
> Great thinking, but It doesn't seem to be possible for branches in forks. The branch from the previous PR doesn't exist in the main repo and I can't target it.

Ah, that's a shame. 

I re-triggered the CI because of a failure and updated it to the latest commit. Looks like there's another release note conflict to resolve though.

kean committed:
Update release notes

kean commented:
@guarani, rebased with `trunk`. If merge is preferred, please let me know.

There is a new "Domains" card that was added to the dashboard in `trunk`. I updated the code to compile but haven't added it to the Personalize screen.

```swift
case .domainsDashboardCard:
    /// TODO
    return DashboardFailureCardCell.self
```

staskus commented:
> @guarani, rebased with `trunk`. If merge is preferred, please let me know.
> 
> There is a new "Domains" card that was added to the dashboard in `trunk`. I updated the code to compile but haven't added it to the Personalize screen.
> 
> ```swift
> case .domainsDashboardCard:
>     /// TODO
>     return DashboardFailureCardCell.self
> ```

@kean thanks for that. There's a new domain card that is in the process of being created and is hidden under the feature flag.

Feel free to proceed with merging this PR, we will make sure to include it in a new personalization view. As I understand, each card will need to be added there manually?

kean commented:
Hey, @staskus! Thanks for the heads up.

Not every card is personalizable, though it looks like the "Domains" card needs to be. @osullivanchris, what do you think?

> As I understand, each card will need to be added there manually?

There are three steps to add the full hide/unhide support for a card (but not everything is merged yet):

- Add a user defaults key to `BlogDashboardPersonalizationService`(more details and rational [here](https://github.com/wordpress-mobile/WordPress-iOS/pull/20365))
- Add the card to `DashboardCard.personalizableCards` and provide a localized title in `BlogDashboardPersonalizationCardCellViewModel` (both added in this PR)
- Add the "Hide this" action to the card's context menu. I'm adding a reusable [`makeHideCardAction`](https://github.com/wordpress-mobile/WordPress-iOS/blob/76b469dd8b41f1c4c4a1febfbc23d828fcad2455/WordPress/Classes/ViewRelated/Blog/Blog%20Dashboard/Helpers/BlogDashboardHelpers.swift) methods to make it easy in  #20384

We could potentially remove one of these steps. The list of personalizable cards could be auto-generated based on whether the card has the associated key, but the order of the cards on the Personalize screen doesn't match the order on the Dashboard, so it needs to be set somewhere. @osullivanchris, can we make the order the same? It makes sense, especially if we eventually add the ability to re-order the cards, which is being discussed.

osullivanchris commented:
> Not every card is personalizable, though it looks like the "Domains" card needs to be. @osullivanchris, what do you think?

Yeah I think the user should be able to hide this if they want to.

> We could potentially remove one of these steps. The list of personalizable cards could be auto-generated based on whether the card has the associated key, but the order of the cards on the Personalize screen doesn't match the order on the Dashboard, so it needs to be set somewhere. @osullivanchris, can we make the order the same? It makes sense, especially if we eventually add the ability to re-order the cards, which is being discussed.

If the order of the menu was automatically derived from the order that cards are displayed no the dashboard, I think that would be great. Whatever order I put in my mockup was arbitrary, just for showing how it would look.

staskus commented:
Thanks for the explanation, @kean!

> Not every card is personalizable, though it looks like the "Domains" card needs to be

Yes, it will probably be. Since we're still creating the card, we'll be happy to try using the flow you described and see if we notice anything that could be improved from the developers' perspective.

> We could potentially remove one of these steps. The list of personalizable cards could be auto-generated based on whether the card has the associated key

That definitely makes sense to me. I don't want to be jumping into this project and stalling the PR. We could make such improvements later if needed.

guarani commented:
> @guarani, rebased with `trunk`. If merge is preferred, please let me know.

👋 Hey @kean, we normally merge when updating a PR from `trunk`. 

I triggered CI on this branch again and enabled auto-merge 👍

hassaanelgarem commented on WordPress/Classes/ViewRelated/Blog/Blog Dashboard/DashboardCard.swift:
```diff
@@ -111,6 +116,15 @@ enum DashboardCard: String, CaseIterable {
     }
 
+    /// A list of cards that can be shown/hidden on a "Personalize Home Tab" screen.
+    static let personalizableCards: [DashboardCard] = [
+        .todaysStats,
+        .draftPosts,
+        .scheduledPosts,
+        .blaze,
+        .prompts
+    ]
+
```
@kean Sorry for being late to the party here, but what do you think about doing something like this?

```suggestion
    // Boolean indicating whether the DashboardCard can be shown/hidden from the "Personalize Home Tab" screen.
    var isPersonalizable: Bool {
        switch self {
        // add a case for each card
        }
    }
    
    static var personalizableCards: [DashboardCard] {
        DashboardCard.allCases.filter({ $0.isPersonalizable })
    }
```

This will help us avoid not adding future cards to the personalize screen if applicable, especially for developers unaware of these new changes.

kean replied:
I like that. The `personalizableCards` were initially defined as an array because the order was important – it didn't match the cards' order on the Dashboard. But I just confirmed late last week that it could and should match.

I'm planning to generate this array based on whether the `BlogDashboardPersonalizationService` defines a key for a card: 

```swift
static var personalizableCards: [DashboardCard] {
    DashboardCard.allCases.filter {
        BlogDashboardPersonalizationService.makeKey(for: $0 != nil)
    }
}
```

The keys are also defined using a `switch`, so the compiler will surface it to any developers adding new cards.

hassaanelgarem replied:
This will achieve the same thing I had in mind 👍 
The only issue is we'll have to make `BlogDashboardPersonalizationService.makeKey` public for this reason alone. It might make more sense to move `personalizableCards` to `BlogDashboardPersonalizationService` at that point. This way, we'll have all the personalization logic in one place. Just something to think about 👍

hassaanelgarem replied:
My personal recommendation would be to change `makeKey()` into a computed property `DashboardCard.personalizeKey` defined in a private extension in `BlogDashboardPersonalizationService.swift`. This will enhance the verbosity of the code.
And move `personalizableCards` to `BlogDashboardPersonalizationService`.

kean replied:
I opened a draft PR with this change: #20466.

> It might make more sense to move personalizableCards to BlogDashboardPersonalizationService at that point. 

Makes total sense, thanks!
